perf: Amortize clearing unsorted cache entries (Juno genesis fix)

blazeroni opened the pull request:
This change fixes a bounty by the Juno team.  Juno's invariant checks took 10 hours during their most recent chain halt. This PR cuts that down to 30 seconds.  See https://github.com/CosmosContracts/bounties#improve-speed-of-invariant-checks.

The root problem is deep in the `can-withdraw` invariant check, which calls this repeatedly: https://github.com/cosmos/cosmos-sdk/blob/main/x/distribution/keeper/store.go#L337.  Iterators have a chain of parents and in this case creates an iterator from the `cachekv` store.  For the genesis file, it has a cache of 500,000+ unsorted entries, which are sorted as strings here: https://github.com/cosmos/cosmos-sdk/blob/main/store/cachekv/store.go#L314.  Each delegation from `can-withdraw` uses this cache and many of the cache checks miss or are a very small range.  This means very few entries get removed from the unsorted cache and they have to be re-sorted on the next call.  With a full cache it takes about 180ms on my machine to sort them.

This change introduce a minimum number of entries that will get processed and removed from the unsorted list. It's set at the same value that directs the code to sort them in the first place.  This ensures the unsorted values get removed in a relative short amount of time, and amortizes the cost to ensure an individual check does not have to process the entire cache.

## Benchmarks
On running the benchmarks included in this change produces:
```shell
name                    old time/op    new time/op    delta
LargeUnsortedMisses-32     21.2s ± 9%      0.0s ± 1%   -99.91%  (p=0.000 n=20+17)

name                    old alloc/op   new alloc/op   delta
LargeUnsortedMisses-32    1.64GB ± 0%    0.00GB ± 0%   -99.83%  (p=0.000 n=19+19)

name                    old allocs/op  new allocs/op  delta
LargeUnsortedMisses-32     20.0k ± 0%     41.1k ± 0%  +105.23%  (p=0.000 n=19+20)
```

## Invariant checks results
This is what the invariant checks for Juno look like with this change (on a Hetzner AX101):

```shell
INF starting node with ABCI Tendermint in-process
4:11PM INF Starting multiAppConn service impl=multiAppConn module=proxy
4:11PM INF Starting localClient service connection=query impl=localClient module=abci-client
4:11PM INF Starting localClient service connection=snapshot impl=localClient module=abci-client
4:11PM INF Starting localClient service connection=mempool impl=localClient module=abci-client
4:11PM INF Starting localClient service connection=consensus impl=localClient module=abci-client
4:11PM INF Starting EventBus service impl=EventBus module=events
4:11PM INF Starting PubSub service impl=PubSub module=pubsub
4:11PM INF Starting IndexerService service impl=IndexerService module=txindex
4:11PM INF ABCI Handshake App Info hash= height=0 module=consensus protocol-version=0 software-version=v9.0.0-36-g8fd6f16
4:11PM INF ABCI Replay Blocks appHeight=0 module=consensus stateHeight=0 storeHeight=0
4:12PM INF asserting crisis invariants inv=1/11 module=x/crisis name=gov/module-account
4:12PM INF asserting crisis invariants inv=2/11 module=x/crisis name=distribution/nonnegative-outstanding
4:12PM INF asserting crisis invariants inv=3/11 module=x/crisis name=distribution/can-withdraw
4:12PM INF asserting crisis invariants inv=4/11 module=x/crisis name=distribution/reference-count
4:12PM INF asserting crisis invariants inv=5/11 module=x/crisis name=distribution/module-account
4:12PM INF asserting crisis invariants inv=6/11 module=x/crisis name=bank/nonnegative-outstanding
4:12PM INF asserting crisis invariants inv=7/11 module=x/crisis name=bank/total-supply
4:12PM INF asserting crisis invariants inv=8/11 module=x/crisis name=staking/module-accounts
4:12PM INF asserting crisis invariants inv=9/11 module=x/crisis name=staking/nonnegative-power
4:12PM INF asserting crisis invariants inv=10/11 module=x/crisis name=staking/positive-delegation
4:12PM INF asserting crisis invariants inv=11/11 module=x/crisis name=staking/delegator-shares
4:12PM INF asserted all invariants duration=28383.559601 height=4136532 module=x/crisis
```

## Alternatives
There is another PR which fixes this problem for the Juno genesis file https://github.com/cosmos/cosmos-sdk/pull/12886. However, because of its concurrent nature, it happens to hit a large range relatively early, clearing the unsorted entries and allowing the rest of the checks to not sort it.

blazeroni commented:
I've updated the PR to a realistic implementation.  It adds a module-internal interface which allows the values to be cached.

blazeroni commented:
I dug deeper and found the root problem.  Replaced most of the code for a real fix.

blazeroni commented:
@odeke-em not sure if you've started looking at this, but the latest code is much different than the previous version. Just a heads up.

blazeroni commented:
Added a benchmark which illustrates the problem.  Finishes in a few seconds with the changes in the PR.  Takes minutes without it.

odeke-em commented on store/cachekv/store.go:
```diff
@@ -325,6 +327,15 @@ func (store *Store) dirtyItems(start, end []byte) {
 		startIndex = 0
 	}
 
+	// since we took the time to sort the cache, we should use that effort
+	// we store at least THRESHOLD values -- cost of storing all is amortized across multiple calls
+	if endIndex-startIndex < THRESHOLD {
+		endIndex = math.MinInt(startIndex+THRESHOLD, len(strL)-1)
+		if endIndex-startIndex < THRESHOLD {
+			startIndex = math.MaxInt(endIndex-THRESHOLD, 0)
+		}
+	}
```
Could you please explain this code more? I don't follow why specifically the span has to fit in < THRESHOLD and what this condition is? Thank you.

blazeroni replied:
The intention is that we actually remove some values from the unsorted cache instead of only a few or possibly none.  This code will extend start/end to at least THRESHOLD values.  If it's already above THRESHOLD, no changes are made.

I've updated the comment here, hopefully it's a bit clearer.  Let me know if you think it needs additional explanation.

odeke-em replied:
Thank you!

odeke-em reviewed (changes requested):
Thank you fro this change @blazeroni and great to meet you! I've added some initial comments but one of them will help to avoid polluting benchmarks. I encourage you to run the benchmarks before and after this change, on a quiet computer that doesn't have a web browser running nor heavy programs, and you can run the benchmarks by doing this

### Before administering the code change in store/cachekv/store.go
```shell
go test -run=^$ -bench=LargeUnsortedMisses -benchmem=true -count=20 > before.txt
```

### After administering the code change in store/cachekv/store.go
```shell
go test -run=^$ -bench=LargeUnsortedMisses -benchmem=true -count=20 > after.txt
```

### Compare the results
To compare the results please firstly install benchstat `go get 
 -u golang.org/x/perf/cmd/benchstat`
```shell
benchstat before.txt after.txt
```

then paste the results into your commit message to clearly exhibit the performance differences

blazeroni commented:
@odeke-em thanks for the feedback!  I've updated the benchmark and added it as a comment.  I had to reduce the size of the original benchmark (from 200k to 10k unsorted values) since the before benchmark was crashing with values much higher than what I used.  Regardless, the difference between before/after is still extreme.

joeabbey commented:
Impressive results.  I've measured similar duration for crisis invariants on similar hardware.  Moreover since this PR changes CacheKV performance, the whole startup from a large `genesis.json` improves.

Without this change the invariants took roughly 14 hours to complete, are there was a stage of replaying blocks and other initialization that took roughly two hours.  Validators were reporting 16+ hours to begin dialing peers.

With this change the entire initialization process completes in around 5 minutes.

kocubinski reviewed (approved):
Reading through the cachekv iterator implementation this seems like an obvious improvement.  Great work!  

If I'm understanding right the only additional side effect here is moving items from the unsorted cache into the sorted cache up to the size of THRESHOLD (1024).  This is a perf improvement, but especially so when `dirtyItems` is called under the following circumstances:

- There are a very large number of unsorted, dirty cache items
- `dirtyItems` is called many times (many iterators are created)
- the range of said iterators is quite small.

This addresses the last point by not wasting the CPU cycles we've already spent in sorting.  Since `memIterator` only ever enumerates the `sortedCache MemDB` I can't see how this would be harmful.  

Does my analysis sound consistent with yours?  I'm tentatively approving, but want to make sure on the same page. 

Can you offer any help in obtaining dataset you used, and some instructions to run so that I can see the benchmarks myself?

blazeroni commented:
@kocubinski yes, your analysis matches my own.

To test/verify this, I used a modified version of the latest Juno production release, available here: https://github.com/blazeroni/juno-bounty/commits/v9.0.0-invariants.  That version depends on a modified v0.45.6 Cosmos SDK, with the only change being the changes in this PR.  Instructions to obtain the Juno genesis file are here: https://github.com/CosmosContracts/incident-response/blob/main/28-July-22/genesis.md.  You'll just need to install, init, replace genesis, and start.

Cheers!

odeke-em reviewed (commented):
I slept on it and just re-read the code and now it makes sense to me. Thank you @blazeroni! Let me make an update of the commit message so as to post the benchmarks and ensure they are seen.

odeke-em reviewed (approved):
LGTM!

alexanderbez commented on store/cachekv/store.go:
```diff
@@ -280,6 +281,7 @@ const (
 
 // Constructs a slice of dirty items, to use w/ memIterator.
 func (store *Store) dirtyItems(start, end []byte) {
+	const THRESHOLD = 1024
```
can we move this outside the method and give it a better name?

alexanderbez replied:
Also explain the reasoning behind 1024 in a godoc comment :)

blazeroni replied:
Sure, though 1024 pre-dates this change and apparently is a magic number: https://github.com/cosmos/cosmos-sdk/pull/10026#discussion_r697965203

It's re-used in this PR since it made sense (to me) that the minimum number of elements removed matches the threshold to sort.

blazeroni replied:
@alexanderbez moved and renamed.  Since it's magic and documented where it's used, I haven't added a godoc comment.  Open to suggestions.

alexanderbez commented:
@marbar3778 do we want this backported?

marbar3778 commented:
> @marbar3778 do we want this backported?

would be nice. I cant see why its not backwards compatible
